### PR TITLE
[Prototype] [Core] Make expensive subpackage imports dynamic.

### LINKED ISSUE #27606: [2.0][nightly-test] many-tasks regression
### What happened + What you expected to happen

many_tasks test regression: 
https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_Ueg7eY6NXWCWW5cmifgvagVF?command-history-section=event_log&drivers-section=deployments passing
https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_wnF3jdENJ3rDBEiJA3E9BsrU?command-history-section=command_history failing

### Versions / Dependencies

master / releases/2.0.0

### Reproduction script

N/A

### Issue Severity

High: It blocks me from completing my task.

### LINKED ISSUE #29557: [Core] Eagerly importing non-core subpackages slows down Python worker startup
Certain Ray subpackages are expensive to import, either due to their size, dependencies, or their import-time logic that must be executed. E.g. a change in a Datasets import caused a regression for the many_tasks nightly test, despite many_tasks only using Ray Core. However, we want subpackages like `data` and `workflow` to be accessible via attributes on a simple `import ray` import:
```python
import ray
ds = ray.data.read_parquet(...)
```

## Possible Solution

One solution is to delay the import of these expensive subpackages until attribute access, e.g. none of the Datasets code will be run until `ray.data` is accessed.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Certain Ray subpackages are expensive to import, either due to their size, dependencies, or their import-time logic that must be executed. E.g. a change in a Datasets import caused a regression for the `many_tasks` nightly test, despite `many_tasks` only using Ray Core.

This PR delays the import of these expensive subpackages until attribute access, e.g. none of the Datasets code will be run until `ray.data` is accessed.

## Related issue number

Closes https://github.com/ray-project/ray/issues/27606, closes #29557

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/__init__.py
Comment (clarkzinzow) [inline python/ray/__init__.py:273]: Note that, similar to an object's `__getattr__` method, this `__getattr__` will only be called if normal attribute looking (e.g. via `module.__getattribute__`) fails, so this will not run for any of the above subpackages that are eagerly imported and should only run for the packages in `_subpackages` (or non-existent module attributes, which will error anyway).

File: python/ray/__init__.py
Comment (clarkzinzow) [inline python/ray/__init__.py:272]: Note that, similar to your typical `from ray import data` statement, imports are cached, so this will translate to a dictionary lookup after the subpackage has been imported once.

File: python/ray/__init__.py
Comment (clarkzinzow) [inline python/ray/__init__.py:267]: We could implement something similar for Python 3.6 via a "replace module with class hack":
```python
class module(types.ModuleType):
    __all__ = __all__

    def __init__(self, orig_module: types.ModuleType):
        self._orig_module = orig_module

    def __getattr__(self, name: str):
        try:
            return getattr(name, self._orig_module)
        except AttributeError as e:
            if name in _subpackages:
                return importlib.import_module("." + name, self._orig_module.__name__)
            raise e from None


sys.modules[__name__] = module(sys.modules[__name__])
```

But IMO this is more hacky than its worth since we're going to be dropping Python 3.6 support in the near-term.

**Comment by rkooo567** [comment]
This one seems to be pretty good? What was the original reason why we decided to defer this PR?

**Comment by rkooo567** [comment]
I'd love to merge it. What's the main concern of this approach?

**Comment by clarkzinzow** [comment]
@rkooo567 it was originally deferred because the `many_tasks` perf gain wasn't very significant (10-15% IIRC) and there are still some odd CI failures to investigate.

File: python/ray/__init__.py
Comment (rkooo567) [inline python/ray/__init__.py:172]: Just OOC, why did it work before?

File: python/ray/__init__.py
Comment (rkooo567) [reply python/ray/__init__.py:272]: Is there a way to test it?

Comment (rkooo567) [reply python/ray/__init__.py:272]: Also if the import fails here, is the error message equivalent to when import <module> failed?

File: python/ray/__init__.py
Comment (rkooo567) [inline python/ray/__init__.py:273]: Can we add a test for this behavior?

File: python/ray/__init__.py
Comment (rkooo567) [inline python/ray/__init__.py:278]: maybe just leave it as it is? A little concern there might have been a reason why we didn't del... haha

**Review by rkooo567 — APPROVED** [review]
LGTM. Maybe we can just add minimal tests? Also the approach seems reasonable. Is there any concern on top of your head doing this?

File: python/ray/__init__.py
Comment (clarkzinzow) [reply python/ray/__init__.py:272]: I believe it should be equivalent, I can check.

File: python/ray/__init__.py
Comment (clarkzinzow) [reply python/ray/__init__.py:278]: We previously weren't importing `sys` in this module, so this shouldn't be a change in behavior. `sys` import was added here: https://github.com/ray-project/ray/pull/27658/files#diff-f95026a08bcb464b58b036437876716d21d3b8630e61258303bcd5384d1d707cR4

File: python/ray/__init__.py
Comment (clarkzinzow) [reply python/ray/__init__.py:273]: Hmm this one might be interesting to test... I suppose we can check that `module.__getattribute__` returns the expected module but errors for `__getattr__` for the top-level subpackages that we define and are eagerly imported?

File: python/ray/__init__.py
Comment (clarkzinzow) [reply python/ray/__init__.py:173]: Not sure 🧐

File: python/ray/tests/test_runtime_env_working_dir_4.py
Comment (rkooo567) [reply python/ray/tests/test_runtime_env_working_dir_4.py:153]: why do we need this change in this PR?

File: python/ray/tests/test_top_level_api.py
Comment (rkooo567) [inline python/ray/tests/test_top_level_api.py:57]: love the test!

**Review by rkooo567 — COMMENTED** [review]
Will merge after a couple runs of many_Tasks. Please **do not** merge latest master @clarkzinzow. I'd like to check if this fixes the issue without @jiaodong'sPR that will be merged soon.

File: python/ray/tests/test_runtime_env_working_dir_4.py
Comment (clarkzinzow) [reply python/ray/tests/test_runtime_env_working_dir_4.py:153]: The test fails otherwise, where the timeout error is never raised

**Comment by clarkzinzow** [comment]
@rkooo567 sounds good, although before merging we should double-check the failing `test_object_directory_failure` test, as well as the tweak that I made to the runtime env test.

**Comment by rkooo567** [comment]
promising sign btw. It passes the test https://buildkite.com/ray-project/release-tests-pr/builds/18731#0183f7a5-b88f-4269-b1ba-b03fdf4f9c8e. I am rerunning it

**Comment by clarkzinzow** [comment]
@rkooo567 No spooky CI failures in the most recent run, worker registration timeout adjustment for the runtime env test appears to have done the trick.

**Comment by clarkzinzow** [comment]
@rkooo567 All currently failing tests were flaky in master at the time of the run, do you think this is good to merge?

**Comment by clarkzinzow** [comment]
Tests are looking good, merging!

**Pull request merged**
